### PR TITLE
fix: retain original array after defaultsDeep #27240

### LINKED ISSUE #27103: Empty returned `specPattern` doesn't yield error if `specPattern` was originally an array
### Current behavior

Setting `specPattern` to an empty array `[]` within `setupNodeEvents() { .. }` makes Cypress correctly yield a `no spec files` error. However, if `specPattern` was originally configured as an array, this is no longer the case.

### Desired behavior

For Cypress to yield the above-mentioned error no matter if `specPattern` was originally specified as a string or an array of strings.

### Test code to reproduce

https://github.com/badeball/reproducible-issues/tree/master/cypress/empty-spec-pattern

### Cypress Version

12.15.0

### Node version

v18.8.0

### Operating System

Arch Linux

### Debug Logs

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes https://github.com/cypress-io/cypress/issues/27103

**note** remaking - original PR https://github.com/cypress-io/cypress/pull/27312 got stuck on red on CI, I could not get it to pass. New branch is fine 🤷 

### Additional details
`What is affected by this change?`
Previously the defaultsDeep function was merging/combining arrays. This PR will fix that issue and restore any arrays modified by the `setupNodeEvents` function.

### Steps to test
To test, copy the cypress.config.js from the `reproduction` section of the issue and configure for e2e testing. Then run cypress using `yarn cypress open --dev --config-file cypress.config.js`. After setting any key within the config object to any array value (within the setupNodeEvents function), it should reflect as is in the final config.

### How has the user experience changed?
User will get the exact config they supply in the `setupNodeEvents` function.

### PR Tasks

- [doubt] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by lmiller1990** [comment]
@NiharPhansalkar https://github.com/cypress-io/cypress/pull/27240 kept failing - I made a new branch and it passes fine, I am not sure why. I've seen this happen before -- some CI weirdness.

Can you approve this? I left an approval on your original PR, but had to remake it. You'll still get credit, your commits are here.

**Comment by NiharPhansalkar** [comment]
Sure will approve!

**Review by NiharPhansalkar — APPROVED** [review]

**Review by cacieprins — APPROVED** [review]

**Review by AtofStryker — APPROVED** [review]

File: packages/app/cypress/e2e/subscriptions/specChange-subscription.cy.ts [hunk 1]
```diff
@@ -444,6 +444,40 @@ e2e: {
 
       cy.get('[data-cy="file-match-indicator"]', { timeout: 7500 })
       .should('contain', '3 matches')
+
+      // Regession for https://github.com/cypress-io/cypress/issues/27103
```
Comment (AtofStryker) [inline packages/app/cypress/e2e/subscriptions/specChange-subscription.cy.ts]:
```suggestion
      // Rregession for https://github.com/cypress-io/cypress/issues/27103
```

File: cli/CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,10 @@ _Released 3/21/2024_
 - Fixed an issue where Cypress was not executing beyond the first spec in `cypress run` for versions of Firefox 124 and up. Fixes [#29172](https://github.com/cypress-io/cypress/issues/29172).
 - Fixed an issue blurring shadow dom elements. Fixed in [#29125](https://github.com/cypress-io/cypress/pull/29125).
 
+**Bugfixes:**
+
+- Fixed a bug where fields using arrays in `cypress.config` are not correctly processed. Fixes [#27103](https://github.com/cypress-io/cypress/issues/27103). Fixed in [#27312](https://github.com/cypress-io/cypress/pull/27240).
```
Comment (AtofStryker) [inline cli/CHANGELOG.md]:
```suggestion
- Fixed a bug where fields using arrays in `cypress.config` are not correctly processed. Fixes [#27103](https://github.com/cypress-io/cypress/issues/27103). Fixed in [#27312](https://github.com/cypress-io/cypress/pull/27312).
```

**Pull request merged**